wasm.md

```diff
@@ -1011,6 +1016,52 @@ The maximum of table size is 2^32 bytes.
     rule #maxTableSize()  => 4294967296
 ```
 
+### Data Segments
+
+Memories can be initialized with data, specified as a list of bytes together with an offset.
+The `data` initializer simply puts these bytes into the specified memory, starting at the offset.
+
+```k
+    syntax Instr ::= Data
+    syntax Data ::= "(" "data" MemId DataOffset DataStrings ")"
+                  | "(" "data"       DataOffset DataStrings ")"
+                  |     "data" "{" MemId DataStrings "}"
+ // ----------------------------------------------------
+    // Default to memory 0.
+    rule <k> ( data       OFFSET STRINGS ) => ( data 0 OFFSET STRINGS ) ... </k>
+    rule <k> ( data MEMID    IS:Instr STRINGS ) => IS ~> data { MEMID STRINGS } ... </k>
+    rule <k> ( data MEMID (offset IS) STRINGS ) => IS ~> data { MEMID STRINGS } ... </k>
+
+    // For now, deal only with memory 0.
+    rule <k> data { MEMIDX STRING } => . ... </k>
+         <valstack> < i32 > OFFSET : STACK => STACK </valstack>
+         <memIndices> MEMIDX |-> ADDR </memIndices>
+         <memInst>
+           <mAddr> ADDR </mAddr>
+           <mdata>   DATA
+                  => #clearRange(DATA, OFFSET, OFFSET +Int #dataStringsLength(STRING) -Int 1) [ OFFSET := #dataStrings2int(STRING)]
+           </mdata>
+           ...
+         </memInst>
+
+    syntax Int ::= #lengthDataPages        ( Data ) [function]
+                 | Int "/ceilInt" Int          [function]
+ // -----------------------------------------------------
+    rule #lengthDataPages((data  _:MemId _ SS)) => #dataStringsLength(SS) /ceilInt #pageSize()
+    rule #lengthDataPages((data          _ SS)) => #dataStringsLength(SS) /ceilInt #pageSize()
+    rule #lengthDataPages( data { _        SS}) => #dataStringsLength(SS) /ceilInt #pageSize()
+
+    rule I1 /ceilInt I2 => (I1 /Int I2) +Int #if I1 modInt I2 ==Int 0 #then 0 #else 1 #fi
+```
+
+The `data` instruction takes an offset, which is an instruction.
+This is not optional.
+
+```k
+    syntax DataOffset ::= "(" "offset" Instr ")"
+                        | Instr
```

Comment:
Hmmm, this looks weird to me. Why can `DataOffset` be an `Instr`?

Reply:
Yeah, it seems a bit redundant, but it's explicitly allowed (and occurs in the official tests):

> As an abbreviation, a single instruction may occur in place of the offset:
> instr ≡ ‘( ‘offset’ instr ‘)’

These are both official tests:
```
  (data (offset (i32.const 0)))
  (data (i32.const 0))
```

Reply:
In fact, that's true not just for `data`, but for `elem` as well, so it should probably just be named `offset` or something.